nixos/modules/services/monitoring/cockpit.nix

```diff
@@ -0,0 +1,225 @@
+    systemd.sockets."cockpit" = {
+      unitConfig = {
+        Description="Cockpit Web Service Socket";
+        Documentation="man:cockpit-ws(8)";
+        Wants="cockpit-motd.service";
+      };
+      socketConfig = {
+        ListenStream=cfg.port;
+        ExecStartPost= [
+          "-${cfg.package}/share/cockpit/motd/update-motd \"\" localhost"
+          "-${pkgs.coreutils}/bin/ln -snf active.motd /run/cockpit/motd"
+        ];
+        ExecStopPost="-${pkgs.coreutils}/bin/ln -snf inactive.motd /run/cockpit/motd";
+      };
+      wantedBy = [ "sockets.target" ];
+    };
+    # Translation from $out/lib/systemd/system/cockpit.service
+    systemd.services."cockpit" = {
+      description = "Cockpit Web Service";
+      documentation = [ "man:cockpit-ws(8)" ];
+      restartIfChanged = true;
+      path = with pkgs; [ coreutils cfg.package ];
+      requires = [ "cockpit.socket" "cockpit-wsinstance-http.socket" "cockpit-wsinstance-https-factory.socket" ];
+      after = [ "cockpit-wsinstance-http.socket" "cockpit-wsinstance-https-factory.socket" ];
+      environment = {
+        G_MESSAGES_DEBUG = "cockpit-ws,cockpit-bridge";
+        # upstream says that systemd ≥ 241 sets this automatically
+        # RUNTIME_DIRECTORY = "/run/cockpit/tls";
+      };
+      serviceConfig = {
+        RuntimeDirectory="cockpit/tls";
+        ExecStartPre=[
+          # cockpit-tls runs in a more constrained environment, these + means that these commands
+          # will run with full privilege instead of inside that constrained environment
+          "+${pkgs.coreutils}/bin/mkdir -p $RUNTIME_DIRECTORY"
+          "+${cfg.package}/libexec/cockpit-certificate-ensure --for-cockpit-tls"
+        ];
+        ExecStart="${cfg.package}/libexec/cockpit-tls";
+        User="root";
+        Group="";
+        NoNewPrivileges=true;
+        ProtectSystem="strict";
+        ProtectHome=true;
+        PrivateTmp=true;
+        PrivateDevices=true;
+        ProtectKernelTunables=true;
+        RestrictAddressFamilies= [ "AF_UNIX" "AF_INET" "AF_INET6" ];
+        MemoryDenyWriteExecute=true;
+      };
+    };
+
+    # Translation from $out/lib/systemd/system/cockpit-motd.service
+    systemd.services."cockpit-motd" = {
+      path = with pkgs; [ gnused nettools ];
+      serviceConfig = {
+        Type = "oneshot";
+        ExecStart = "${cfg.package}/share/cockpit/motd/update-motd";
+      };
+      description = "Cockpit motd updater service";
+      documentation = [ "man:cockpit-ws(8)" ];
+      wants = [ "network-online.target" ];
+      after = [ "network-online.target" "cockpit.socket" ];
+    };
+
+    systemd.tmpfiles.rules = [ # From $out/lib/tmpfiles.d/cockpit-tmpfiles.conf
+      "C /run/cockpit/inactive.motd 0640 root root - ${cfg.package}/share/cockpit/motd/inactive.motd"
+      "f /run/cockpit/active.motd   0640 root root -"
+      "L+ /run/cockpit/motd - - - - inactive.motd"
+    ];
+
+    # cockpit login doesn't work without this
```

Comment:
I assume cockpit requires some pam service or so. (Not really familiar with pam) The comment should reflect that instead of "it happens to work with that".

Reply:
It provides the PAM modules automagically that are accessible via `/run/current-system/sw/lib/security/pam_cockpit_cert.so` so modules works out of the box by the logic upstream and Nix stdenv provides.

But it only logs in users with the pam service.

Reply:
BTW I removed that comment